### PR TITLE
fix(helm): Use `semver-coerced` for determining latest release

### LINKED ISSUE #16223: Helm datasource refuses whole repository when a single version fails to parse
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

32.97

### Please select which platform you are using if self-hosting.

_No response_

### If you're self-hosting Renovate, tell us what version of the platform you run.

_No response_

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

The Helm datasource fails to load the `aws-eks` repository, conducing my own investigations I found that one of the releases was [published as `0.1`](https://github.com/aws/eks-charts/blame/98e0a8f15aa6480cf71730efc0a4a3f4abe8a394/stable/aws-sigv4-proxy-admission-controller/Chart.yaml#L4) instead of `0.1.0`, this throws an exception when sorting which discard the whole repository.

I added a testcase with the index.yaml file showing this failure here: https://github.com/meroje/renovate/commit/cfa0d190884dc8689afe2be55b349e66e4a683d0 ([tests run](https://github.com/Meroje/renovate/runs/7045151896?check_suite_focus=true#step:7:137))

I think `versioning/semver-coerced` could be used, not sure if it should be. Let me know if you'd like a patch doing this.

### Relevant debug logs

<details><summary>Logs</summary>

```
 WARN: Failed to parse index.yaml from helm repository (repository=xx/xx)
       "helmRepository": "https://aws.github.io/eks-charts"
DEBUG:  (repository=xx/xx)
DEBUG: Couldn't get index.yaml file from https://aws.github.io/eks-charts (repository=xx/xx)
DEBUG: Failed to look up dependency aws-load-balancer-controller (repository=xx/xx, packageFile=xxx/aws-load-balancer-controller/Chart.yaml, dependency=aws-load-balancer-controller)
```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction repository in the bug description

### PR DESCRIPTION
**Pull request by zharinov**
## Changes

- Switch from pure semver comparison to our `semver-coerced` one

## Context

- Closes: #16223
## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by rarkins — DISMISSED** [review]

File: lib/modules/datasource/helm/index.ts [hunk 1]
```diff
@@ -63,9 +63,9 @@ export class HelmDatasource extends Datasource {
       }
       const result: HelmRepositoryData = {};
       for (const [name, releases] of Object.entries(doc.entries)) {
-        const latestRelease = releases.sort((r0, r1) =>
-          gt(r0.version, r1.version) ? -1 : 1
-        )[0];
+        const [latestRelease] = releases.sort((r0, r1) =>
+          semverCoerced.api.isGreaterThan(r0.version, r1.version) ? -1 : 1
```
Comment (viceice) [inline lib/modules/datasource/helm/index.ts:67]: we do have a sort function on our versioning which should be used instead 😉

Comment (Meroje) [reply lib/modules/datasource/helm/index.ts:67]: It only sorts versions but here we’re working on releases with a version property

**Review by Meroje — COMMENTED** [review]
You’ll need to coerce the version when building the result array too, would you like me to send in the failing test about this issue to protect regressions ?

**Comment by zharinov** [comment]
I think we should address the root cause as separate issue: why our `helm` versioning scheme implementation doesn't support values like `1.0`?

**Comment by rarkins** [comment]
Was it created for custom versions or for custom constraint syntax?

**Comment by zharinov** [comment]
I have no idea. And at this point, I think we have to get rid of sort and use the latest element from array instead.

File: lib/modules/datasource/helm/index.ts [hunk 2]
```diff
@@ -63,9 +62,11 @@ export class HelmDatasource extends Datasource {
       }
       const result: HelmRepositoryData = {};
       for (const [name, releases] of Object.entries(doc.entries)) {
-        const latestRelease = releases.sort((r0, r1) =>
-          gt(r0.version, r1.version) ? -1 : 1
-        )[0];
+        const [latestRelease] = releases
+          .filter(({ version }) => helmVersioning.api.isValid(version))
+          .sort((r0, r1) =>
+            helmVersioning.api.isGreaterThan(r0.version, r1.version) ? -1 : 1
```
Comment (viceice) [inline lib/modules/datasource/helm/index.ts:68]:
```suggestion
            helmVersioning.api.sortVersions(r0.version, r1.version)
```
as i mentioned before: use `sortVersions`. it's there for exactly this purpose.

**Review by viceice — CHANGES_REQUESTED** [review]

**Comment by zharinov** [comment]
I think this PR fixes the original problem while mostly preserving current behavior. I'm not sure if I should care about versions like that `1.0` which aren't supported by our Helm versioning scheme.

File: lib/modules/datasource/helm/index.ts [hunk 3]
```diff
@@ -63,9 +62,11 @@ export class HelmDatasource extends Datasource {
...
       const result: HelmRepositoryData = {};
       for (const [name, releases] of Object.entries(doc.entries)) {
-        const latestRelease = releases.sort((r0, r1) =>
-          gt(r0.version, r1.version) ? -1 : 1
-        )[0];
+        const [latestRelease] = releases
+          .filter(({ version }) => helmVersioning.api.isValid(version))
+          .sort((r0, r1) =>
+            helmVersioning.api.sortVersions(r0.version, r1.version)
+          );
```
Comment (viceice) [inline lib/modules/datasource/helm/index.ts:69]: we should have a fallback if the filtered list is empty . then last release should be used.

Comment (viceice) [reply lib/modules/datasource/helm/index.ts:69]:
```suggestion
          );

  if(!latestRelease) {
    latestRelease = releases[release.length - 1];
  }
```

**Review by viceice — APPROVED** [review]
probably needs a test for coverage

**Pull request merged**